unix/x0vncserver/XSelection.cxx

```diff
@@ -0,0 +1,144 @@
+/* Copyright (C) 2024 Gaurav Ujjwal.  All Rights Reserved.
+ *
+ * This is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this software; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307,
+ * USA.
+ */
+
+#include "XSelection.h"
+#include <X11/Xatom.h>
+#include <rfb/LogWriter.h>
+#include <rfb/util.h>
+
+static rfb::LogWriter vlog("XSelection");
+
+XSelection::XSelection(Display* dpy_, XSelectionHandler* handler_)
+    : TXWindow(dpy_, 1, 1, nullptr), handler(handler_)
+{
+  xaTransferProperty = XInternAtom(dpy, "SelectionTransferProperty", False);
+  xaTimestampProperty = XInternAtom(dpy, "TimestampAccessProperty", False);
+  addEventMask(PropertyChangeMask); // Required for PropertyNotify events
+}
+
+static Bool PropertyEventMatcher(Display* /* dpy */, XEvent* ev, XPointer prop)
+{
+  if (ev->type == PropertyNotify && ev->xproperty.atom == *((Atom*)prop))
+    return True;
+  else
+    return False;
+}
+
+Time XSelection::getXServerTime()
+{
+  XEvent ev;
+  uint8_t data = 0;
+
+  // Trigger a PropertyNotify event to extract server time
+  XChangeProperty(dpy, win(), xaTimestampProperty, XA_STRING, 8,
+                  PropModeReplace, &data, sizeof(data));
+  XIfEvent(dpy, &ev, &PropertyEventMatcher, (XPointer)&xaTimestampProperty);
+  return ev.xproperty.time;
```

Comment:
Why not just `CurrentTime`?

Reply:
I was following the advice from https://x.org/releases/X11R7.7/doc/xorg-docs/icccm/icccm.html#Peer_to_Peer_Communication_by_Means_of_Selections : 

`Clients attempting to acquire a selection must set the time value of the SetSelectionOwner request to the timestamp of the event triggering the acquisition attempt, not to CurrentTime. A zero-length append to a property is a way to obtain a timestamp for this purpose; the timestamp is in the corresponding PropertyNotify event. `

I did use `CurrentTime` initially, and it worked without an issue. But since this time is also used [here](https://github.com/TigerVNC/tigervnc/blob/2fe9dca45f7d0a519fef5053ba61e2db7c1ff2b1/unix/tx/TXWindow.cxx#L472-L475) , I changed it to current implementation just to be safe. 
I can switch it back if you want.

Reply:
Ah, we need it for `TIMESTAMP`. That complicates things.

*Normally*, you take the event time that caused us to grab the selection. E.g. when the user pressed "Ctrl+C". Unfortunately, we are not normal. We get our events from RFB, not from X11, so we don't have a readily available time. "Right now" is as good as it gets, and that is precisely what "CurrentTime" means.

But that only solves the call for `XSetSelectionOwner()`. So I think your suggestion is the best approach.